node/sync/locators/src/block_locators.rs

```diff
@@ -189,27 +193,44 @@ impl<N: Network> BlockLocators<N> {
         // Ensure the block checkpoints are well-formed.
         let last_checkpoint_height = Self::check_block_checkpoints(checkpoints)?;
 
-        // Ensure the `last_recent_height` is at or above `last_checkpoint_height - NUM_RECENT_BLOCKS`.
-        let threshold = last_checkpoint_height.saturating_sub(NUM_RECENT_BLOCKS as u32);
-        if last_recent_height < threshold {
-            bail!("Recent height ({last_recent_height}) cannot be below checkpoint threshold ({threshold})")
-        }
-
-        // If the `last_recent_height` is below NUM_RECENT_BLOCKS, ensure the genesis hash matches in both maps.
-        if last_recent_height < NUM_RECENT_BLOCKS as u32
-            && recents.get(&0).copied().unwrap_or_default() != checkpoints.get(&0).copied().unwrap_or_default()
+        // Ensure that `last_checkpoint_height` is
+        // the largest multiple of `CHECKPOINT_INTERVAL` that does not exceed `last_recent_height`.
+        // That is, we must have
+        // `last_checkpoint_height <= last_recent_height < last_checkpoint_height + CHECKPOINT_INTERVAL`.
+        // Although we do not expect to run out of `u32` for block heights,
+        // `last_checkpoint_height` is an untrusted value that may come from a faulty validator,
+        // and thus we use a saturating addition;
+        // only a faulty validator would send block locators with such high block heights,
+        // under the assumption that the blockchain is always well below the `u32` limit for heights.
+        if !(last_checkpoint_height..last_checkpoint_height.saturating_add(CHECKPOINT_INTERVAL))
+            .contains(&last_recent_height)
         {
-            bail!("Recent genesis hash and checkpoint genesis hash mismatch at height {last_recent_height}")
+            bail!(
+                "Last checkpoint height ({last_checkpoint_height}) is not the largest multiple of ".to_string()
+                    + "{CHECKPOINT_INTERVAL} that does not exceed the last recent height ({last_recent_height})"
```

Comment:
nit: Why is a string concatenation required here?

```suggestion
                "Last checkpoint height ({last_checkpoint_height}) is not the largest multiple of {CHECKPOINT_INTERVAL} that does not exceed the last recent height ({last_recent_height})"
```

Reply:
I thought we might get a clippy complaint for exceeding 120 columns?

Reply:
You can make it more concise w/ something like:
``` 
bail!("Invalid checkpoint ({last_checkpoint_height}): must be the largest multiple of {CHECKPOINT_INTERVAL} ≤ {last_recent_height}");
```

Should be fine as long as clippy does not complain. Alternatively you can use the "\" character to split the lines. 

For example:
```
bail!(
    "Checkpoint height ({last_checkpoint_height}) must be the largest multiple of \
     {CHECKPOINT_INTERVAL} not exceeding recent height ({last_recent_height})"
);
```

Reply:
Yes, I should have used `\`. Pushed.